restmapper + discovery: add context-aware APIs

pohly opened the pull request:
#### What type of PR is this?

/kind feature

#### What this PR does / why we need it:

The main purpose is to replace context.TODO with a context provided by the caller. A secondary purpose is to enable contextual logging.

#### Which issue(s) this PR fixes:

Fixes https://github.com/kubernetes/kubernetes/issues/110810

https://github.com/kubernetes/kubernetes/issues/119813

#### Special notes for your reviewer:

Modifying the existing interfaces and APIs would have a big impact on the ecosystem. This is a no-go. Instead, the following approach was taken:

- All interfaces get duplicated in a *WithContext variant where the methods also have a *WithContext suffix and the ctx parameter. All methods are treated this way except for obvious local get methods (like RESTClient) because it cannot be ruled out entirely that some implementation may need a context.

- Implementations of these interfaces implement both method variants, which is possible because the method names are different. The old methods are implemented as thin wrappers around the updated code which is now the body of the new methods or shared helpers. In some cases there is additional overhead (type checks, potentially additional allocations) when using the old methods.

- `*2Context` helpers bridge from the old to the new interfaces. They try a type cast first. Because the in-tree implementations implement both, they can be used directly. For other implementations wrappers are used.

- None of old APIs and interfaces are marked as deprecated. Instead  doc comments inform consumers about the new and better alternatives.    This approach avoids causing turmoil in the ecosystem when linters    pick up on a formal deprecation and developers act on those linter    warnings without properly thinking about the implications.
    
  "// Contextual logging: " is used as placeholder for "//logcheck:context".      Once all relevant consumers have been updated to the new API,       a search/replace can enable the logcheck linter to warn about       usage of the old API.
    
  Implementations also get marked this way even if probably nothing  ever calls them directly because it shows which code, at least   theoretically, could get removed.

- Existing unit tests do not get updated to the new APIs. This gives us unit test coverage of the old and new API because the old APIs call the new ones.

- In-tree consumers will be updated in follow-up PRs. This is likely to be a longer process. Because of the deprecation comment, `hack/golangci-lint.sh -n` can be used to find code which needs to be updated.

#### Does this PR introduce a user-facing change?
```release-note
client-go now has context-aware APIs for REST mapping and discovery. This addresses a long-standing problem that the implementations under the hood made blocking API calls with `context.TODO`. Consumers of client-go are encouraged to switch to the new APIs and therefore they get marked as `Deprecated`. However, there is no plan to ever remove the old APIs.
```

/cc @ardaguclu @sttts

pohly committed:
restmapper + discovery: add context-aware APIs

The main purpose is to replace context.TODO with a context provided by the
caller. A secondary purpose is to enable contextual logging.

Modifying the existing interfaces and APIs would have a big impact on the
ecosystem. This is a no-go. Instead, the following approach was taken:

- All interfaces get duplicated in a *WithContext variant where the methods
  also have a *WithContext suffix and the ctx parameter. All methods are
  treated this way except for obvious local get methods (like RESTClient)
  because it cannot be ruled out entirely that some implementation may
  need a context.

- Implementations of these interfaces implement both method variants
  which is possible because the method names are different.
  The old methods are implemented as thin wrappers around the updated
  code which is now the body of the new methods or shared helpers.
  In some cases there is additional overhead (type checks, potentially
  additional allocations) when using the old methods.

- To*WithContext helpers bridge from the old to the new interfaces. They
  try a type cast first. Because the in-tree implementations implement
  both, they can be used directly. For other implementations wrappers
  are used.

- None of old APIs and interfaces are marked as deprecated. Instead
  doc comments inform consumers about the new and better alternatives.
  This approach avoids causing turmoil in the ecosystem when linters
  pick up on a formal deprecation and developers act on those linter
  warnings without properly thinking about the implications.

  "// Contextual logging: " is used as placeholder for "//logcheck:context".
  Once all relevant consumers have been updated to the new API,
  a search/replace can enable the logcheck linter to warn about
  usage of the old API.

  Implementations also get marked this way even if probably nothing
  ever calls them directly because it shows which code, at least
  theoretically, could get removed.

- Existing unit tests do not get updated to the new APIs. This gives
  us unit test coverage of the old and new API because the old
  APIs call the new ones.

- In-tree consumers will be updated in follow-up PRs. This is likely
  to be a longer process. Because of the deprecation comment,
  `hack/golangci-lint.sh -n` can be used to find code which needs
  to be updated.

pohly commented on staging/src/k8s.io/apimachinery/pkg/api/meta/interfaces.go:
```diff
@@ -133,11 +135,112 @@ type RESTMapper interface {
+// unique across groups.
+type RESTMapperWithContext interface {
+	// KindFor takes a partial resource and returns the single match.  Returns an error if there are multiple matches
+	KindForWithContext(ctx context.Context, resource schema.GroupVersionResource) (schema.GroupVersionKind, error)
```
I am using the `*WithContext` pattern consistently because that's what we settled on in Kubernetes. However, it leads to long names which in this case are also a bit odd.

In Go, `*Context` is used more often than `*WithContext`:
```
# e.g. database/sql/sql.go:func (db *DB) PingContext(ctx context.Context) error {
$ grep -r 'func.*Context(ctx' /nvme/gopath/go/src/ | grep -v WithContext | wc -l
55
# e.g. net/http/request.go:func NewRequestWithContext(ctx context.Context, method, url string, body io.Reader) (*Request, error)
$ grep -r 'func.*WithContext(ctx' /nvme/gopath/go/src/ | wc -l
5
```

Should we follow that practice and use the shorter `*Context` pattern?

Then this becomes:
```golang
type RESTMapperContext interface {
    KindForContext(ctx context.Context, resource schema.GroupVersionResource) (schema.GroupVersionKind, error)
```

ardaguclu replied:
Although `*Context` is used far more and makes sense, `RESTMapperContext` may mislead people about it behaves as a Context rather than having a Context.

ardaguclu reviewed (commented):
Without any in-depth code review, overall method seems to be the most plausible one for least disruption and makes sense to me. 
We may need to ask additional opinions from;
openapi/discovery
@seans3 @Jefftree 
restmapper
@deads2k  @liggitt

Jefftree commented:
/triage accepted

pohly commented:
> Adding label do-not-merge/contains-merge-commits because PR contains merge commits, which are not allowed in this repository.

This is intentional. https://github.com/kubernetes/kubernetes/pull/129109 has to be merged first. In the meantime, I am keeping a merge commit here as a reminder and to support a workflow based on `git rebase --keep-base --rebase-merges -i --autosquash` when maintaining several related PRs.

pohly commented:
/test pull-kubernetes-apidiff

pohly commented on staging/src/k8s.io/apimachinery/pkg/api/meta/interfaces.go:
```diff
@@ -133,11 +135,112 @@ type RESTMapper interface {
+	if m, ok := m.(RESTMapperWithContext); ok {
+		return m
+	}
+	return &restMapperWrapper{
```
I wasn't entirely sure whether this should return a value or a pointer to satisfy the interface. Both works, so I benchmarked it. Mem allocs are the same (one alloc/call), but returning a value is 10% slower when calling one method.

What I then benchmarked is whether pointer vs value receiver makes a difference. Converting from value receiver (what I had before) to pointer receiver reduced call overhead by 6%. That's because with a value receiver, Go has to auto-generate a variant with pointer receiver and calls the value receiver through that.

That can be seen in a debugger (call stack) and when setting breakpoints:
```
(dlv) b restMapperWrapper.KindForWithContext
Command failed: Location "restMapperWrapper.KindForWithContext" ambiguous: k8s.io/apimachinery/pkg/api/meta.restMapperWrapper.KindForWithContext, k8s.io/apimachinery/pkg/api/meta.(*restMapperWrapper).KindForWithContext…
```

Conventional wisdom is to define types with value receiver because those can be called also on unmutable instances, making them more flexible.

But for types which will only ever be used via a pointer, I think pointer receiver is better for the reasons above (small performance difference, easier to debug).

I'll change to that in the entire PR.

pohly commented:
/test pull-kubernetes-apidiff

I am done with updating the code and doing a full pass over it. This is ready for review.

/assign @soltysh

pohly commented:
/test pull-kubernetes-e2e-gce pull-kubernetes-node-e2e-containerd

Test cluster failed to come up.

soltysh commented on staging/src/k8s.io/apimachinery/pkg/api/meta/interfaces.go:
```diff
@@ -109,7 +111,7 @@ type RESTMapping struct {
 // unique across groups.
 //
-// TODO: split into sub-interfaces
+// Deprecated: use RESTMapperWithContext instead.
```
This applies to other files as well. I remember we agreed NOT to deprecate these, but rather put a message like:
```
// Contextual logging: RESTMapperWithContext should be used instead of RESTMapper in code which supports contextual logging.
```

pohly replied:
That was for functions where the original version is functionally sufficient for its intended purpose.

Here we have a case of an interface that looks like it works locally (no context) but under the hood often must call blocking functions which really should be called with a context if the user of this interface has one. I think this is a case where we should nudge the ecosystem towards updating their code by declaring the old functions as deprecated.

soltysh reviewed (commented):
I left a few comments, but the major ones are:
1. I believe we agreed to NOT do deprecation on methods w/o context.
2. What should we do with tests in `staging/src/k8s.io/apimachinery/pkg/api/meta/` directory, currently they only work with non-context variants. Should we duplicate them for context variants? 

The rest seems reasonable, but I will want to do another pass.

pohly commented on staging/src/k8s.io/apimachinery/pkg/api/meta/priority.go:
```diff
@@ -30,11 +31,14 @@ const (
 var (
-	_ ResettableRESTMapper = PriorityRESTMapper{}
+	_ ResettableRESTMapper            = PriorityRESTMapper{}
+	_ ResettableRESTMapperWithContext = PriorityRESTMapperWithContext{}
```
Done.

pohly replied:
Note that I decided to keep using `PriorityRESTMapperWithContext` as value: that's because `ToPriorityRESTMapperWithContext` returns the actual struct, not an interface, and the struct is small enough that passing it by value should be better, in particular because all callers immediately discard it again.

pohly commented on staging/src/k8s.io/apimachinery/pkg/api/meta/restmapper.go:
```diff
@@ -72,6 +73,7 @@ func (m *DefaultRESTMapper) String() string {
 }
 
 var _ RESTMapper = &DefaultRESTMapper{}
+var _ RESTMapperWithContext = &DefaultRESTMapper{}
```
Done.

pohly commented:
/remove-lifecycle stale

harche commented:
LGTM from my side, but @soltysh do the changes look okay to you?

soltysh reviewed (commented):
There are a few minor nits here, but I believe the two major comments from last time hasn't been addressed, there are:
1. I believe we agreed to NOT do deprecation on methods w/o context. Especially in the light of discussion around https://github.com/kubernetes/kubernetes/pull/136455, I'd say deprecation only on interfaces/types, but leave methods untouched, at minimum. I'm still not sure about interfaces/types, but I'd say doing this on types/interfaces might have a smaller impact than modifying all methods. Especially that there's a lot of them. 
2. What should we do with tests in staging/src/k8s.io/apimachinery/pkg/api/meta/ directory, currently they only work with non-context variants. Should we duplicate them for context variants?

pohly commented on staging/src/k8s.io/client-go/restmapper/category_expansion.go:
```diff
@@ -17,48 +17,113 @@ limitations under the License.
+	}
+	if c == nil {
+		return nil
+	}
```
Cat jumped on my keyboard! Wait... which cat? Never mind, fixed.

soltysh reviewed (commented):
There are some minor nits, but based on my review this 
/lgtm

pohly committed:
client-go: expose all discovery methods

The underlying implementation has both. Exposing that in the result of
Discovery() makes it easier for consumers to call the WithContext methods
because they don't need to wrap with ToDiscoveryInterfaceWithContext().

This is a breaking Go API change because tests which implement
the interface have to be updated.

sttts commented:
Codex xhigh review:
```
  - P1: A canceled context can initialize and cache an incomplete RESTMapper.
    /private/tmp/k8s-pr-129109/staging/src/k8s.io/client-go/restmapper/discovery.go:187 ignores err whenever gs and rs are non-nil, and /private/tmp/k8s-pr-129109/staging/src/k8s.io/client-go/restmapper/discovery.go:261 then caches the mapper at /private/tmp/k8s-pr-129109/staging/src/k8s.io/client-go/restmapper/discovery.go:266. With the new context-aware path, per-GV resource requests record
    any error, including context.Canceled / DeadlineExceeded, as failedGroups in /private/tmp/k8s-pr-129109/staging/src/k8s.io/client-go/discovery/discovery_client.go:1081, then return partial results plus ErrGroupDiscoveryFailed at /private/tmp/k8s-pr-129109/staging/src/k8s.io/client-go/discovery/discovery_client.go:972. That means a deadline during first mapper initialization can permanently
    cache a mapper missing resources until reset. Context errors should abort mapper initialization instead of being treated as tolerable partial discovery.

  - P1: Memory discovery cache stores per-call context failures as durable cache entries.
    In /private/tmp/k8s-pr-129109/staging/src/k8s.io/client-go/discovery/cached/memory/memcache.go:387, refreshLocked stores every ServerResourcesForGroupVersionWithContext error into rl, then marks the whole cache valid at /private/tmp/k8s-pr-129109/staging/src/k8s.io/client-go/discovery/cached/memory/memcache.go:407. Later reads only retry isTransientError entries at /private/tmp/k8s-pr-
    129109/staging/src/k8s.io/client-go/discovery/cached/memory/memcache.go:120; context cancellation/deadline errors are not transient, so one canceled refresh can poison discovery until Invalidate. Context-derived errors should not be cached as resource state.

  - P2: OpenAPI caches permanently cache first-call context cancellation.
    /private/tmp/k8s-pr-129109/staging/src/k8s.io/client-go/openapi/cached/client.go:73 uses sync.Once; if the first PathsWithContext(ctx) fails because ctx is canceled, /private/tmp/k8s-pr-129109/staging/src/k8s.io/client-go/openapi/cached/client.go:76 stores that error forever. Similarly, /private/tmp/k8s-pr-129109/staging/src/k8s.io/client-go/openapi/cached/groupversion.go:67 stores
    SchemaWithContext errors per content type. A per-call context failure should be retryable by a later call with a live context.

  - P2: Retry loop does not stop on context cancellation wrapped as group discovery failure.
    /private/tmp/k8s-pr-129109/staging/src/k8s.io/client-go/discovery/discovery_client.go:1234 retries every ErrGroupDiscoveryFailed until maxRetries, with no way to distinguish context cancellation from “API groups changed under us.” Since /private/tmp/k8s-pr-129109/staging/src/k8s.io/client-go/discovery/discovery_client.go:1087 now puts context errors inside ErrGroupDiscoveryFailed, canceled
    calls can do extra discovery rounds instead of returning promptly.
```

pohly commented on staging/src/k8s.io/client-go/restmapper/discovery.go:
```diff
@@ -135,17 +168,23 @@ func NewDiscoveryRESTMapper(groupResources []*APIGroupResources) meta.RESTMapper
+// GetAPIGroupResourcesWithContext uses the provided discovery client to gather
+// discovery information and populate a slice of APIGroupResources.
+func GetAPIGroupResourcesWithContext(ctx context.Context, cl discovery.DiscoveryInterfaceWithContext) ([]*APIGroupResources, error) {
+	gs, rs, err := cl.ServerGroupsAndResourcesWithContext(ctx)
```
https://github.com/kubernetes/kubernetes/pull/129109#issuecomment-4659847623:

```
  - P1: A canceled context can initialize and cache an incomplete RESTMapper.
    /private/tmp/k8s-pr-129109/staging/src/k8s.io/client-go/restmapper/discovery.go:187 ignores err whenever gs and rs are non-nil, and /private/tmp/k8s-pr-129109/staging/src/k8s.io/client-go/restmapper/discovery.go:261 then caches the mapper at /private/tmp/k8s-pr-129109/staging/src/k8s.io/client-go/restmapper/discovery.go:266. With the new context-aware path, per-GV resource requests record
    any error, including context.Canceled / DeadlineExceeded, as failedGroups in /private/tmp/k8s-pr-129109/staging/src/k8s.io/client-go/discovery/discovery_client.go:1081, then return partial results plus ErrGroupDiscoveryFailed at /private/tmp/k8s-pr-129109/staging/src/k8s.io/client-go/discovery/discovery_client.go:972. That means a deadline during first mapper initialization can permanently
    cache a mapper missing resources until reset. Context errors should abort mapper initialization instead of being treated as tolerable partial discovery.
```

Codex isn't wrong here, but this isn't new. The "// TODO track the errors and update callers to handle partial errors." below tracks this.

One could argue that cancellation is a new class of failures and should get special treatment, but I don't think so.

sttts replied:
The point is to stop early when ctx is already done. But could argue that `ServerGroupsAndResourcesWithContext` should then do that and return `rs+gs==nil`.

pohly commented on staging/src/k8s.io/client-go/discovery/cached/memory/memcache.go:
```diff
@@ -275,16 +382,16 @@ func (d *memCacheClient) refreshLocked() error {
+				defer utilruntime.HandleCrashWithContext(ctx)
 
-				r, err := d.serverResourcesForGroupVersion(gv)
+				r, err := d.serverResourcesForGroupVersion(ctx, gv)
```
https://github.com/kubernetes/kubernetes/pull/129109#issuecomment-4659847623

```
  - P1: Memory discovery cache stores per-call context failures as durable cache entries.
    In /private/tmp/k8s-pr-129109/staging/src/k8s.io/client-go/discovery/cached/memory/memcache.go:387, refreshLocked stores every ServerResourcesForGroupVersionWithContext error into rl, then marks the whole cache valid at /private/tmp/k8s-pr-129109/staging/src/k8s.io/client-go/discovery/cached/memory/memcache.go:407. Later reads only retry isTransientError entries at /private/tmp/k8s-pr-
    129109/staging/src/k8s.io/client-go/discovery/cached/memory/memcache.go:120; context cancellation/deadline errors are not transient, so one canceled refresh can poison discovery until Invalidate. Context-derived errors should not be cached as resource state.
```

So the complaint here is that the error gets logged  and stored instead of handling it. Again, that's how the existing code worked. I'm not sure what's correct and whether we need to do better. Let's see...

The context in this case is passed in by the user in ServerResourcesForGroupVersionWithContext. Not returning the error here causes it to continue and check the cached error:

https://github.com/pohly/kubernetes/blob/5e7b6fedc07a6932eea6b9f0fecf2b35178fba66/staging/src/k8s.io/client-go/discovery/cached/memory/memcache.go#L115-L136

The ServerResourcesForGroupVersionWithContext with a cancelled context returns the same error either way because it's stored in `cachedVal.err`.

It's a second call to ServerResourcesForGroupVersionWithContext with a fresh context where things get interesting: it will not call `d.refreshLocked` again because the cache is considered valid. It'll then check the cached error and *not* recognize it as transient, thus returning it without the retry through `d.serverResourcesForGroupVersion(ctx, groupVersion)`. This is the "can poison discovery until Invalidate" part.

IMHO the problem is `isTransientError`: instead of retrying for a few known transient errors, the cache should skip retrying for known permanent errors. But changing that is a bigger effort.

I extended the unit tests to cover this and fixed the failing test by treating context cancellation as transient in `isTransientError`.

While at it I also added a unit test that verifies that the conversion to the WithContext variant works.

pohly commented on staging/src/k8s.io/client-go/openapi/cached/client.go:
```diff
@@ -17,34 +17,67 @@ limitations under the License.
+func (c *Client) PathsWithContext(ctx context.Context) (map[string]openapi.GroupVersionWithContext, error) {
 	c.once.Do(func() {
-		uncached, err := c.delegate.Paths()
+		uncached, err := c.delegate.PathsWithContext(ctx)
```
https://github.com/kubernetes/kubernetes/pull/129109#issuecomment-4659847623

```
  - P2: OpenAPI caches permanently cache first-call context cancellation.
    /private/tmp/k8s-pr-129109/staging/src/k8s.io/client-go/openapi/cached/client.go:73 uses sync.Once; if the first PathsWithContext(ctx) fails because ctx is canceled, /private/tmp/k8s-pr-129109/staging/src/k8s.io/client-go/openapi/cached/client.go:76 stores that error forever. Similarly, /private/tmp/k8s-pr-129109/staging/src/k8s.io/client-go/openapi/cached/groupversion.go:67 stores
    SchemaWithContext errors per content type. A per-call context failure should be retryable by a later call with a live context.
```

The existing code lacks automatic retries. Any transient error, not just context cancellation, is permanent. As this is the  existing semantic I prefer to not change this?

sttts replied:
the point is that ctx could be cancelled on entering the func. One could (in general) check for it and avoid side-effects. No strong opinion.

pohly replied:
Ack => https://github.com/kubernetes/kubernetes/pull/129109/changes/c3f63fc258b73e487da7ca84c40be43fc4c5f4dd.

pohly commented on staging/src/k8s.io/client-go/discovery/discovery_client.go:
```diff
@@ -648,23 +1114,57 @@ func (d *DiscoveryClient) ServerPreferredResources() ([]*metav1.APIResourceList,
+// ServerPreferredResourcesWithContext returns the supported resources with the version preferred by the
+// server.
+func (d *DiscoveryClient) ServerPreferredResourcesWithContext(ctx context.Context) ([]*metav1.APIResourceList, error) {
+	_, rs, err := withRetries(defaultRetries, func() ([]*metav1.APIGroup, []*metav1.APIResourceList, error) {
```
https://github.com/kubernetes/kubernetes/pull/129109#issuecomment-4659847623

```
  - P2: Retry loop does not stop on context cancellation wrapped as group discovery failure.
    /private/tmp/k8s-pr-129109/staging/src/k8s.io/client-go/discovery/discovery_client.go:1234 retries every ErrGroupDiscoveryFailed until maxRetries, with no way to distinguish context cancellation from “API groups changed under us.” Since /private/tmp/k8s-pr-129109/staging/src/k8s.io/client-go/discovery/discovery_client.go:1087 now puts context errors inside ErrGroupDiscoveryFailed, canceled
    calls can do extra discovery rounds instead of returning promptly.`
```

Line number is a bit off, but this refers to `withRetries`. A valid complaint. I changed `withRetries` to check for context cancellation errors.

pohly commented:
/test pull-kubernetes-unit

TestTunnelingHandler_UpstreamSPDYServerErrorPropagated - unrelated.

sttts commented on staging/src/k8s.io/client-go/restmapper/shortcut.go:
here `e.expandResourceShortcut(resource)` got lost

pohly replied:
It gets called by `KindsForWithContext` below. All we need to do here is inject `context.Background()`, the actual implementation of `KindsFor*` is in `KindsForWithContext`.

sttts replied:
then you want `e.KindsForWithContext`, no?

pohly replied:
Ah, now I see it. Yes, you are right. Thanks, fixed!

pohly committed:
restmapper + discovery: early return in context-aware APIs

The rationale for explicit cancellation checks is that these functions have
side effects that can be avoided when checking upfront. These side effects can
still occur when cancellation happens while the functions are executing, but
"already canceled" is common enough that it deserves special handling.

sttts commented:
/lgtm
/approve

pohly commented on pkg/kubelet/clustertrustbundle/clustertrustbundle_manager_test.go:
This file needs SIG Node approval.

/cc @SergeyKanzhelev @mrunalp @haircommander

SergeyKanzhelev replied:
/approve

pohly commented:
/test pull-kubernetes-integration

`TestPodGroupScheduling/gang_pods_are_unschedulable_due_to_lack_of_space,_then_scheduled_when_minCount_is_decreased_(TopologyAwareWorkloadScheduling_enabled:_false)` - unrelated flake?

/test pull-kubernetes-node-e2e-containerd

`Device Plugin Failures Pod Status [FeatureGate:ResourceHealthStatus] will report a Healthy and then Unhealthy single device in the pod status [Beta] [FeatureGate:DynamicResourceAllocation] [NodeConformance]` - also unrelated?

pohly commented:
/skip

pohly commented:
/test pull-kubernetes-e2e-gce

Cluster didn't start.
